### DISCUSSION TITLE
How to reach the template directory from a router.py in a sub diretory?

### DISCUSSION THREAD

**Discussion by waterchinap** [fastapi/fastapi#7033] [Questions] [answered]
I tried to follow the tutoria to form a bigger application
https://fastapi.tiangolo.com/tutorial/bigger-applications/
the project structure is like :
```
.
├── app
│   ├── __init__.py
│   ├── main.py
│   ├── model.py
│   ├── countries.gz
│   └── routers
│         ├── __init__.py
│         ├── items.py
│         └── users.py
│   └── templates
│         ├── layout.html
│         ├── home.html
│         └── world.html
```
main.py:
```python
from fastapi import FastAPI, Request, APIRouter
from .routers import us_router, world_router

app = FastAPI()
# should I add some config here?

app.include_router(us_router.router)
app.include_router(world_router.wrouter)

```

in the /routers/items.py
```python
from fastapi.templating import Jinja2Templates

wrouter = APIRouter()

templates = Jinja2Templates(directory='templates')
# if the router.py is in the same level of templates directory, this will work.
# should I change something here? I tried '../templates', not work.

@wrouter.get("/world", response_class=HTMLResponse)
async def world(request: Request):
    flist = fig.fig_world('total_cases')
    return templates.TemplateResponse('world.html', {'request': request,
                                                     'line_latest': flist[0],
                                                     }
                                      )

```
The problem:
cannot find the world.html, which is in templates directory.
If I use a absolute directory 'home/eric/fastapi/app/templates', it will work.

model.py:
```python
import pandas as pd

df = pd.read_pickle('countries.gz')
# this is very confusing, I tried './countries.gz', not work

......

```
Problem:
cannot find the 'countries.gz' in the same directory. use absolute directory will work.

expected answer:
As a beginner, I am confused on the logic. I would like to know what is the best practice to handle the directories in the fastapi project?

Thanks a lot.

**Comment by includeamin** [comment]
can you put here your executes command?
do you run the command inside `app` directory or outside of that?

**Comment by waterchinap** [comment, marked as answer]
I execute just outside the app directory: uvicorn app.main:app --reload

OH!
change like this will work:
```
templates = Jinja2Templates(directory='./app/templates')
```

Thank you so much, the project root dir is the excute directory, NOT the main.py.directory.

My follow up question is:
Will it be more reasonable if I set the main.py directory as root? where should I config, If I want to do this,

**Comment by Mause** [comment]
You probably want to do something like this:


```python
from os.path import dirname, join

current_dir = dirname(__file__)  # this will be the location of the current .py file
templates = join(current_dir, 'templates')
```

**Comment by waterchinap** [comment]
Thanks!
wonderful!

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:

**Comment by houmie** [comment]
Pretty odd. What do we do with unit tests?

```tests``` folder is at the same level as ```app```  in my case.

I have to do this to make it work: 
```templates = Jinja2Templates(directory="../app/templates")```

But that won't work then when running the actual server outside unit testing. It's a catch 22.  Is there a better way to achieve this? Thanks

**Comment by LaundroMat** [comment]
@houmie What I did is create the templates variable inside a config.py file in my project root. That way, I can import templates from whatever place in the app (sub-routers, main app, tests, ...) and not worry about directory paths.

**Comment by tiangolo** [comment]
@houmie if you still have issues, please create a new issue following the template to be able to understand your problem. :nerd_face: 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by houmie** [comment]
@tiangolo No problem.
I ended up solving this with an absolute path instead of relative path:

**config.py**

```
@lru_cache()
def get_app_root():
    return os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))

```
**dashboard.py**

```
from fastapi.templating import Jinja2Templates
templates = Jinja2Templates(directory=get_app_root() + "/app/templates")

@router_board.get("/", response_class=HTMLResponse)
async def analysis(request: Request, credentials: HTTPBasicCredentials = Depends(security)):
    return templates.TemplateResponse("dashboard.html", {...})
```
